File: packages/python/elizaos/features/basic_capabilities/providers/entities.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("ENTITIES")


async def get_entities_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    room_id = message.room_id
    sender_id = message.entity_id

    sections: list[str] = []
    entity_list: list[dict[str, str]] = []

    if sender_id:
        sender = await runtime.get_entity(sender_id)
        if sender:
            sender_info = {
                "id": str(sender.id),
                "name": sender.name or "Unknown",
                "type": sender.entity_type or "user",
                "role": "sender",
            }
            entity_list.append(sender_info)
            sections.append(
                f"- **{sender.name or 'Unknown'}** (sender): {sender.entity_type or 'user'}"
            )

    if room_id:
        room = await runtime.get_room(room_id)
        if room and room.world_id:
            world = await runtime.get_world(room.world_id)
            if world and world.metadata:
                member_ids = world.metadata.get("members", [])
                roles = world.metadata.get("roles", {})

                for member_id in member_ids:
                    if str(member_id) == str(sender_id):
                        continue

                    entity = await runtime.get_entity(member_id)
                    if entity:
                        role = roles.get(str(member_id), "member")
                        entity_info = {
                            "id": str(entity.id),
                            "name": entity.name or "Unknown",
                            "type": entity.entity_type or "user",
                            "role": role,
                        }
                        entity_list.append(entity_info)
                        sections.append(
                            f"- **{entity.name or 'Unknown'}** ({role}): {entity.entity_type or 'user'}"
                        )

    agent_entity = await runtime.get_entity(runtime.agent_id)
    if agent_entity:
        agent_info = {
            "id": str(agent_entity.id),
            "name": agent_entity.name or runtime.character.name,
            "type": "agent",
            "role": "self",
        }
        entity_list.append(agent_info)
        sections.append(f"- **{agent_entity.name or runtime.character.name}** (self): agent")

    context_text = "# Entities in Context\n" + "\n".join(sections) if sections else ""

    return ProviderResult(
        text=context_text,
        values={
            "entityCount": len(entity_list),
            "hasSender": sender_id is not None,
            "agentId": str(runtime.agent_id),
        },
        data={
            "entities": entity_list,
            "senderId": str(sender_id) if sender_id else None,
            "agentId": str(runtime.agent_id),
        },
    )


entities_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_entities_context,
    dynamic=_spec.get("dynamic", True),
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/entities.rs
//! ENTITIES provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("ENTITIES"));

/// Provider for entity information.
pub struct EntitiesProvider;

#[async_trait]
impl Provider for EntitiesProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let mut sections = Vec::new();
        let mut entity_list: Vec<serde_json::Value> = Vec::new();

        let sender_id = message.entity_id;

        // Get sender entity
        if let Some(sender_id) = sender_id {
            if let Ok(Some(sender)) = runtime.get_entity(sender_id).await {
                let sender_name = sender.name.clone().unwrap_or_else(|| "Unknown".to_string());
                let sender_type = sender
                    .entity_type
                    .clone()
                    .unwrap_or_else(|| "user".to_string());

                entity_list.push(serde_json::json!({
                    "id": sender_id.to_string(),
                    "name": sender_name,
                    "type": sender_type,
                    "role": "sender"
                }));

                sections.push(format!("- **{}** (sender): {}", sender_name, sender_type));
            }
        }

        // Add agent itself
        let agent_id = runtime.agent_id();
        if let Ok(Some(agent)) = runtime.get_entity(agent_id).await {
            let agent_name = agent
                .name
                .clone()
                .unwrap_or_else(|| runtime.character().name.clone());

            entity_list.push(serde_json::json!({
                "id": agent_id.to_string(),
                "name": agent_name,
                "type": "agent",
                "role": "self"
            }));

            sections.push(format!("- **{}** (self): agent", agent_name));
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Entities in Context\n{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("entityCount", entity_list.len() as i64)
            .with_value("hasSender", sender_id.is_some())
            .with_value("agentId", agent_id.to_string())
            .with_data("entities", entity_list)
            .with_data("agentId", agent_id.to_string()))
    }
}
